remixer_storage/src/main/java/com/google/android/libraries/remixer/storage/FirebaseRemoteControllerSyncer.java

```diff
@@ -17,28 +17,105 @@
 import com.google.android.libraries.remixer.sync.SynchronizationMechanism;
+import com.google.firebase.database.ChildEventListener;
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.FirebaseDatabase;
 import java.lang.ref.WeakReference;
+import java.util.List;
+import java.util.Locale;
+import java.util.UUID;
 
 /**
  * A {@link SynchronizationMechanism} that syncs up to a firebase Remote Controller.
  *
  * <p>This synchronization mechanism assumes that the local host is the source of truth, so it does
  * very little in terms of conflict resolution.
- *
- * <p>This implementation currently ignores bringing new values in from the remote controller, that
- * will be done in firebase callbacks in a future PR, so it is a different one that introduces said
- * dependency.
  */
-public class FirebaseRemoteControllerSyncer extends LocalValueSyncing {
+public class FirebaseRemoteControllerSyncer
+    extends LocalValueSyncing implements ChildEventListener {
 
-  WeakReference<Object> context;
+  /**
+   * The reference to the root element for this Remixer Instance in the database
+   */
+  private DatabaseReference reference;
+  /**
+   * The current context that is active on the app (foreground activity).
+   */
+  private WeakReference<Object> context;
 
-  public FirebaseRemoteControllerSyncer() {
+  /**
+   * Whether syncing is enabled or not. All accesses to this field must be synchronized.
+   */
+  private boolean syncing = false;
+
+  /**
+   * The autogenerated id for this instance of Remixer. This is the first 8 characters of a random
+   * UUID that would identify this device.
+   */
+  private String remoteId;
+
+  private static final String PREFERENCES_FILE_NAME = "remixer_firebase";
+  private static final String REMOTE_ID = "remote_id";
+  private static final String REFERENCE_FORMAT = "remixer/%s";
+
+  /**
+   * Initializes a {@code FirebaseRemoteControllerSyncer} instance.
+   *
+   * <p>Uses {@code context} to get
+   */
+  public FirebaseRemoteControllerSyncer(Context context) {
     super();
+    SharedPreferences preferences =
+        context.getSharedPreferences(PREFERENCES_FILE_NAME, Context.MODE_PRIVATE);
+    remoteId = preferences.getString(REMOTE_ID, null);
+    if (remoteId == null) {
+      remoteId = UUID.randomUUID().toString().substring(0, 7);
+      preferences.edit().putString(REMOTE_ID, remoteId).apply();
+    }
+    startSyncing();
+  }
+
+  /**
+   * Starts syncing up to Firebase.
+   *
+   * <p>Since we don't know what the status of the database is right now, we clear the value and
+   * then add everything for the current context, then add this instance as a child event listener.
+   *
+   * <p>That way we get notified of changes to children (individual variables) after the initial
+   * sync.
+   */
+  public synchronized void startSyncing() {
```

Comment:
also curiosu about whether `synchronized` is necessary here and below. Do you expect any of the synchronized methods to be called from different threads? Why?